utlis/function.js

```diff
@@ -0,0 +1,29 @@
+function search(nameFrag, targetArray) {
+  // chicking array and string
+  if (!nameFrag || !targetArray.length) return null;
+
+  return targetArray.filter((obj) => obj.name.includes(nameFrag));
+}
+
+function addToCartList(id, targetArray, wantedArray) {
+  if (
+    typeof id == "undefined" ||
+    !targetArray.length ||
+    !Array.isArray(wantedArray)
+  )
+    return null;
+  let rArray = [...wantedArray];
+  targetArray.forEach((obj) => obj.id == id && rArray.push(obj));
+  return rArray;
+}
+
+function addToItemList(obj, targetArray) {
+  if (!Array.isArray(targetArray)) return null;
+  if (obj.id && obj.name && obj.price) return [...targetArray, obj];
+}
```

Comment:
This function needs to be pure, you are just copying the elements of `targetArray` and adding `obj` to it, you are returning a new array that's fine, BUT this way you are making a shallow copy since the elements of `targetArray` are reference type(objects). 

```suggestion
function addToItemList(obj, targetArray) {
  if (!Array.isArray(targetArray)) return null;
  const deeplyCopiedArray = JSON.parse(JSON.stringify(targetArray))
  deeplyCopiedArray.push({...obj}) // copying obj before appending it
  if (obj.id && obj.name && obj.price) return [...deeplyCopiedArray, copiedObject];
}
```